src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.cs

```diff
@@ -13,15 +13,34 @@
 using Microsoft.CodeAnalysis.SolutionCrawler;
 using Microsoft.VisualStudio.Threading;
 using Roslyn.Utilities;
-using static Microsoft.VisualStudio.Threading.ThreadingTools;
 
 namespace Microsoft.CodeAnalysis.Remote
 {
+
     /// <summary>
     /// Workspace created by the remote host that mirrors the corresponding client workspace.
     /// </summary>
     internal sealed partial class RemoteWorkspace : Workspace
     {
+        public sealed class PinnedSolution : System.IAsyncDisposable
```

Comment:
this portion of the PR is the scariest.  Prior to this PR we had the model where remote services pass a simple async callback to do work to RemoteWorkspace.  RemoteWorkspace itself pins the solution and invokes the simple callback.

this is no longer simple as we have to support streaming and we have to ensure the solution is pinned while that happens.  Once done with the streaming we have to unpin the solution. (or, decrement the ref count on it and unpin iif it goes to zero).